content/blog/2019-12-21-logical-indexing-on-HB/index.md

```diff
@@ -0,0 +1,249 @@
++++
+title = "Logical Tensor Indexing for SPMD Programming"
+[extra]
+katex_enable = true
+latex = true
+[[extra.authors]]
+name = "Edwin Peguero"
++++
+
+My current research explores programming abstractions for an experimental, parallel architecture (codenamed HammerBlade).
+The language feature prototyped in this project aims to simplify *SPMD* programming, wherein all nodes apply the same kernel over different pieces of input data.
+
+# Cross-cutting Concerns in SPMD Programming
+The need to distribute inputs over cores in SPMD programs introduces problems that are orthogonal to the kernel, but that nonetheless affect implementation:
+- **Data format of inputs/outputs**: 
+  Each tensor $A$ must be encoded in a particular **format**, $f_A$, that maps the **logical index** of each tensor element $A\langle x \rangle \langle y \rangle$ to **physical index** $f_A(x,y)$ into array $A'$, s.t., $A\langle x \rangle \langle y \rangle = A'[f_A(x,y)]$.
+
+  Low-level languages like C, require this dimensionality reduction to maximize performance contributors such as data locality and bulk memory allocations/movements.
+  Different formatting choices lead to *different implementations* of the *same kernel*.
+
+  We constrain our prototype to use the **row-major order** format, which sequentially lays out rows of $A$ via $f_A(x,y) = x * stride + y$, where $stride$ indicates the number of elements per row of $A'$.
+  Our results generalize to similar formats (e.g., *col-major order*), and we leave exploration of more exotic formats to future work.
+
+- Caching:
+  Preemptive promotion of required inputs in the memory hierarchy, another form of data locality, often leads to significant performance gains.
+  However, the choice of caching can change data formatting. 
+  For instance, if a **row-major order** sub-matrix is cached from a **row-major order** matrix, its stride will be different.
+  Thus, the choice of caching also leads to *different implementations* of the *same kernel*.
+
+This project implements a formal separation of **logical indexing** from **data formatting** via a simple 2-D stencil algorithm example.
+
+# Stencil Algorithm Using Logical Indexing
+The proposed syntax for logical indexing uses angle brackets `A<x><y>` to distinguish from traditional array indexing `A[x][y]`.
+We demonstrate this language feature by means of a 2-D stencil algorithm that computes 2-D tensor $B\langle x \rangle \langle y \rangle$ from tensor $A$ as the average of $A\langle x \rangle \langle y \rangle$ and its neighbors. 
+
+At the boundaries of $B$ (i.e., $B\langle x \rangle \langle y \rangle$, $B\langle x \rangle \langle y \rangle$, $B\langle x \rangle \langle y \rangle$, $B\langle x \rangle \langle y \rangle$), the algorithm simply assigns the input directly.
+In SPMD style, each core transforms a block of data, denoted $A_tile$ into a block of output, denoted $B_tile$.
```

Comment:
$A_tile$ will render as A(t)ile, as if it were ${A_{t}}ile$. You probably want $A_{\text{tile}}$ or similar.